teps/0126-allow-task-sidecars-to-be-specified-in-pipelinerun.md

```diff
@@ -0,0 +1,123 @@
+---
+status: proposed
+title: Allow Task sidecars to be specified in PipelineRun
+creation-date: '2022-11-11'
+last-updated: '2022-11-11'
+authors:
+- '@michaelsauter'
+collaborators: []
+---
+
+# TEP-0126: Allow Task sidecars to be specified in PipelineRun
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Extend `PipelineRun` to allow authors to specify sidecars of tasks.
+
+## Motivation
+
+Make tasks more reusable.
+
+Tasks are meant to be shared (e.g. through Git or the ArtifactHub) and referenced from pipeline runs. This TEP seeks to make tasks more often reusable by removing the limitation that sidecars are "hardcoded" and cannot be changed by the pipeline run author.
+
+### Goals
+
+* Extend `PipelineRun` to allow to specify all aspects of sidecars, matching how [sidecars are specified in tasks](https://tekton.dev/docs/pipelines/tasks/#specifying-sidecars).
```

Comment:
Why `PipelineRun` directly and not start with a `Pipeline`?
A `Pipeline` can be signed and verified, and it provides the same benefits of Task reusability.

If someone wants to reproduce a test run in their local cluster, having the correct setup in the `Pipeline` would provide a nicer user experience.

Reply:
The problem with `Pipeline` is that it doesn't allow to e.g. specify compute resource (at least not to my understanding).

Some context: We have a YAML file in each repo which specifies a list of Tekton tasks. That file is read by a "pipeline manager" which is triggered when a push happens in the repo, and the "pipeline manager" creates a `PipelineRun` reflecting the tasks in the YAML file in the repo. There never is a `Pipeline` resource in our cluster - the pipeline is defined in code.

This is also the approach that OpenShift pipelines-as-code has taken (to my understanding). In my experience, the distinction between pipeline and pipeline run is a bit odd and even confusing, especially when using Git as the source of truth for the "pipeline definition". And as I wrote at the beginning, a lot of the features we want to expose to users are only available in pipeline runs, hence we need to expose them to the users.